nomad/worker.go

```diff
@@ -70,70 +113,322 @@ type Worker struct {
+		Started:           w.Started,
+		Status:            w.Status,
+		WorkloadStatus:    w.WorkloadStatus,
+	}
+	copy(out.EnabledSchedulers, w.EnabledSchedulers)
+	return out
+}
+
+func (w WorkerInfo) String() string {
+	// lazy implementation of WorkerInfo to string
+	out, _ := json.Marshal(w)
+	return string(out)
+}
+
+func (w *Worker) Info() WorkerInfo {
 	w.pauseLock.Lock()
-	for w.paused {
+	defer w.pauseLock.Unlock()
+	out := WorkerInfo{
+		ID:                w.id,
+		Status:            w.status.String(),
+		WorkloadStatus:    w.workloadStatus.String(),
+		EnabledSchedulers: make([]string, len(w.enabledSchedulers)),
+	}
+	out.Started = w.start
+	copy(out.EnabledSchedulers, w.enabledSchedulers)
+	return out
+}
+
+// ----------------------------------
+//  Pause Implementation
+//    These functions are used to support the worker's pause behaviors.
+// ----------------------------------
+
+func (w *Worker) setPauseFlag(pause bool) {
+	w.pauseLock.Lock()
+	defer w.pauseLock.Unlock()
+	w.pauseFlag = pause
+}
+
+// maybeWait is responsible for making the transition from `pausing`
+// to `paused`, waiting, and then transitioning back to the running
+// values.
+func (w *Worker) maybeWait() {
+	w.pauseLock.Lock()
+	defer w.pauseLock.Unlock()
+
+	if !w.pauseFlag {
+		return
+	}
+
+	w.statusLock.Lock()
+	w.status = WorkerPaused
+	originalWorkloadStatus := w.workloadStatus
+	w.workloadStatus = WorkloadPaused
+	w.logger.Trace("changed workload status", "from", originalWorkloadStatus, "to", w.workloadStatus)
+
+	w.statusLock.Unlock()
+
+	for w.pauseFlag {
 		w.pauseCond.Wait()
 	}
+
+	w.statusLock.Lock()
+
+	w.logger.Trace("changed workload status", "from", w.workloadStatus, "to", originalWorkloadStatus)
+	w.workloadStatus = originalWorkloadStatus
+
+	// only reset the worker status if the worker is not resuming to stop the paused workload.
+	if w.status != WorkerStopping {
+		w.logger.Trace("changed worker status", "from", w.status, "to", WorkerStarted)
+		w.status = WorkerStarted
+	}
+	w.statusLock.Unlock()
+}
+
+// Shutdown is used to signal that the worker should shutdown.
+func (w *Worker) shutdown() {
+	w.pauseLock.Lock()
+	wasPaused := w.pauseFlag
+	w.pauseFlag = false
 	w.pauseLock.Unlock()
+
+	w.logger.Trace("shutdown request received")
+	w.cancelFn()
+	if wasPaused {
+		w.pauseCond.Broadcast()
+	}
+}
+
+// markStopped is used to mark the worker as stopped and should be called in a
+// defer immediately upon entering the run() function.
+func (w *Worker) markStopped() {
+	w.logger.Debug("stopped")
+	w.setStatus(WorkerStopped)
 }
 
+func (w *Worker) workerShuttingDown() bool {
+	select {
+	case <-w.ctx.Done():
+		return true
+	default:
+		return false
+	}
+}
+
+// ----------------------------------
+//  Workload behavior code
+// ----------------------------------
+
 // run is the long-lived goroutine which is used to run the worker
 func (w *Worker) run() {
+	defer func() {
+		w.setWorkloadStatus(WorkloadStopped)
+		w.markStopped()
```

Comment:
I think we missed this one:

```suggestion
		w.setStatuses(WorkerStopped , WorkloadStopped)
```